phase/phase.go

```diff
@@ -117,47 +137,68 @@ func (self *Phase) Options(s godip.Validator, nation godip.Nation) godip.Options
 	return s.Options(self.Parser.Orders(), nation)
 }
 
-func (self *Phase) Messages(s godip.Validator, nation godip.Nation) []string {
-	messages := []string{}
-	if self.Ty == godip.Adjustment {
-		unitsPerNat := map[godip.Nation]int{}
-		scsPerNat := map[godip.Nation]int{}
-		nats := map[godip.Nation]bool{}
-		freeHomeSCsPerNat := map[godip.Nation]int{}
-		for _, unit := range s.Units() {
-			unitsPerNat[unit.Nation] += 1
-			nats[unit.Nation] = true
-		}
-		for _, nat := range s.SupplyCenters() {
-			scsPerNat[nat] += 1
-			nats[nat] = true
-		}
-		for _, nat := range s.Graph().Nations() {
-			for _, sc := range s.Graph().SCs(nat) {
-				if _, _, found := s.Unit(sc); !found {
-					freeHomeSCsPerNat[nat] += 1
+// Returns number of allowed (after considering free owned SCs where builds are allowed considering the validator
+// flags) builds/needed disbands per nation still in the game.
+func (self *Phase) allowedBuildBalance(s godip.Validator) map[godip.Nation]int {
+	unitsPerNat := map[godip.Nation]int{}
+	scsPerNat := map[godip.Nation]int{}
+	nats := map[godip.Nation]bool{}
+	freeSelfHomePerNat := map[godip.Nation]int{}
+	freeAnyHomePerNat := map[godip.Nation]int{}
+	freeAnywherePerNat := map[godip.Nation]int{}
+
+	for _, unit := range s.Units() {
+		unitsPerNat[unit.Nation] += 1
+		nats[unit.Nation] = true
+	}
+	for sc, nat := range s.SupplyCenters() {
+		scsPerNat[nat] += 1
+		nats[nat] = true
+		if _, _, found := s.Unit(sc); !found {
+			if originalOwner := s.Graph().SC(sc); originalOwner == nil {
+				freeAnywherePerNat[nat] += 1
+			} else {
+				if *originalOwner == nat {
+					freeSelfHomePerNat[nat] += 1
+				} else {
+					freeAnyHomePerNat[nat] += 1
 				}
 			}
-			nats[nat] = true
 		}
-		for nat := range nats {
-			delta := scsPerNat[nat] - unitsPerNat[nat]
+	}
```

Comment:
This feels backwards - we're computing all possible responses and then checking the flags to decide which to return.

Reply:
You are correct, but adding that many if statements around the computing was hard to read :/ 

If you give me an easy to read alternative (I hate bugs in this code, it's messy to test and and so on) I'm open to that :D